File: exercise3/src/nav_goal.cpp
#include <ros/ros.h>
#include <move_base_msgs/MoveBaseAction.h>
#include <actionlib/client/simple_action_client.h>

typedef actionlib::SimpleActionClient<move_base_msgs::MoveBaseAction> MoveBaseClient;

int main(int argc, char** argv){
  
  actionlib::SimpleClientGoalState goal_state = actionlib::SimpleClientGoalState::LOST;
  
  ros::init(argc, argv, "nav_goal_cpp");

  //tell the action client that we want to spin a thread by default
  MoveBaseClient ac("move_base", true);

  //wait for the action server to come up
  while(!ac.waitForServer(ros::Duration(2.0))){
    ROS_INFO("Waiting for the move_base action server to come up");
  }

  move_base_msgs::MoveBaseGoal goal;

  //Sending a goal to the robot to move 1 meter forward
  //goal.target_pose.header.frame_id = "base_link";
  //goal.target_pose.header.stamp = ros::Time::now();
  //goal.target_pose.pose.position.x = 1.0;
  //goal.target_pose.pose.orientation.w = 1.0;

  //Sending a goal to the to a certain position in the map
  goal.target_pose.header.frame_id = "map";
  goal.target_pose.header.stamp = ros::Time::now();
  goal.target_pose.pose.position.x = 0.0;
  goal.target_pose.pose.position.y = 0.0;
  goal.target_pose.pose.orientation.w = 1.0;

  ROS_INFO("Sending goal");
  ac.sendGoal(goal);

  while(not (goal_state == actionlib::SimpleClientGoalState::SUCCEEDED)){

	  ac.waitForResult(ros::Duration(2.0));
	  goal_state = ac.getState();
	  //Possible States Are: PENDING, ACTIVE, RECALLED, REJECTED, PREEMPTED, ABORTED, SUCCEEDED, LOST.

	  if( not (goal_state == actionlib::SimpleClientGoalState::SUCCEEDED)){
	    ROS_INFO("The goal has not been reached yet! Checking again in 2s.");
	  }else{
	    ROS_INFO("The goal was reached!");
	  }
  }

  return 0;
}

File: exercise3/scripts/nav_goal.py
#!/usr/bin/env python

import rospy
import actionlib
from actionlib_msgs.msg import GoalStatus
from move_base_msgs.msg import MoveBaseAction, MoveBaseGoal

goal_state = GoalStatus.LOST
goal_points = [(0.7,-0.28,0),(1.25,-1.47,1),(0,-3,1),(-0.8,-1.6,2),(1.9,-3.2,1)]

rospy.init_node('map_navigation', anonymous=False)

ac = actionlib.SimpleActionClient("move_base", MoveBaseAction)

while(not ac.wait_for_server(rospy.Duration.from_sec(2.0))):
              rospy.loginfo("Waiting for the move_base action server to come up")

for point in goal_points:
    goal_state = GoalStatus.LOST
    goal = MoveBaseGoal()

#Sending a goal to the to a certain position in the map
    goal.target_pose.header.frame_id = "map"
    goal.target_pose.header.stamp = rospy.Time.now()
    goal.target_pose.pose.position.x = point[0]
    goal.target_pose.pose.position.y = point[1]
    goal.target_pose.pose.orientation.w = point[2]

    rospy.loginfo("Sending goal")
    ac.send_goal(goal)

    while (not goal_state == GoalStatus.SUCCEEDED):

        ac.wait_for_result(rospy.Duration(2))
        goal_state = ac.get_state()
        #Possible States Are: PENDING, ACTIVE, RECALLED, REJECTED, PREEMPTED, ABORTED, SUCCEEDED, LOST.

        if goal_state == GoalStatus.SUCCEEDED:
            rospy.loginfo("The goal was reached!")
        elif goal_state == GoalStatus.REJECTED or goal_state == GoalStatus.ABORTED or goal_state == GoalStatus.LOST:
            rospy.loginfo("Goal could not be reached!")
            break
        else:
            rospy.loginfo("Searching for the goal!")


GoalStatus.SUCCEEDED
